src/System.Management.Automation/engine/debugger/debugger.cs

```diff
@@ -4081,7 +4081,24 @@ internal void TraceVariableSet(string varName, object value)
             // because 'ToStringParser' would iterate through the enumerator to get the individual elements, which will
             // make irreversible changes to the enumerator.
             bool isValueAnIEnumerator = PSObject.Base(value) is IEnumerator;
-            string valAsString = isValueAnIEnumerator ? typeof(IEnumerator).Name : PSObject.ToStringParser(_context, value);
+            string valAsString = string.Empty;
+            if (isValueAnIEnumerator)
+            {
+                valAsString = typeof(IEnumerator).Name;
+            }
+            else
+            {
+                try
+                {
+                    valAsString = PSObject.ToStringParser(_context, value);
+                }
+                catch
+                {
+                    // ToStringParser() can throw if the value is already disposed.
+                    valAsString = "null";
+                }
+            }
```

Comment:
Hmm, I think the exception should be handled within `PSObject.ToStringParser`, where the `ToString()` method is called.

Reply:
In debug/tracer we can mask any exception from ToString() without concern.
In engine code I am not sure that we can blindly do the same. 
Docs say that best practice is "avoid exceptions in ToString() methods". So we should consider this as a bug and we shouldn't mask this.

Reply:
@daxian-dbw Could you please continue?

Reply:
@daxian-dbw Could you please make a conclusion?

Reply:
> Docs say that best practice is "avoid exceptions in ToString() methods". So we should consider this as a bug and we shouldn't mask this.

I agree with this, but I think it means we should fix the exception in `PSObject.ToStringParser()`

Reply:
The PR is only for the trace code. For ToStringParser() we could have an issue with repo steps to discuss. It is not clear how many code paths we need to fix there if need.